plonk-core/src/lookup/multiset.rs

```diff
@@ -0,0 +1,525 @@
+    pub fn push(&mut self, value: F) {
+        self.0.push(value)
+    }
+
+    /// Extendes values onto the end of the Multiset
+    pub fn extend<T>(&mut self, iter: T)
+    where
+        T: IntoIterator<Item = F>,
+    {
+        self.0.extend(iter)
+    }
+
+    /// Fetches last element in MultiSet.
+    /// Returns None if there are no elements in the MultiSet.
+    pub fn last(&self) -> Option<&F> {
+        self.0.last()
+    }
+
+    /// Returns the cardinality of the multiset
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns whether or not the multiset is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns the position of the element in the Multiset.
+    /// Returns None if the element is not found.
+    pub fn position(&self, element: &F) -> Option<usize> {
+        self.0.iter().position(move |x| x == element)
+    }
+
+    /// Concatenates and sorts two Multisets together.
+    ///
+    /// All elements of f must exist in t.
+    /// From the Plookup paper, if we have t: {1,2,4,3} and f: {2,3,4,1}.
+    /// We combine the multisets together and sort their elements together.
+    /// The final MultiSet will look as follows, s: {1,1,2,2,3,3,4,4}.
+    /// Splits a multiset into alternating halves of the same length (if even).
+    pub fn sorted_halve(&self, f: &Self) -> Result<(Self, Self), Error> {
+        let mut counters: HashMap<F, usize> = HashMap::new();
+
+        // Insert elemnts on of t in sorted struct
+        for element in &self.0 {
+            match counters.get_mut(element) {
+                Some(v) => *v += 1,
+                _ => {
+                    counters.insert(*element, 1);
+                }
+            };
+        }
+
+        // Insert elements on of f in sorted struct + check they are in t
+        for element in &f.0 {
+            match counters.get_mut(element) {
+                Some(entry) => *entry += 1,
+                _ => return Err(Error::ElementNotIndexed),
+            }
+        }
+
+        let n_elems = self.len() + f.len();
+        let half_len = n_elems / 2;
+        let mut evens = Vec::with_capacity(half_len + (n_elems % 2));
+        let mut odds = Vec::with_capacity(half_len);
+        let mut parity = 0;
+        self.0.iter().for_each(|&elem| {
+            let count = counters.get_mut(&elem).unwrap();
+            let half_count = *count / 2;
+            evens.extend(vec![elem; half_count]);
+            odds.extend(vec![elem; half_count]);
+            // if odd count => add extra element to corresponding
+            // vec and flip prev_parity
+            if *count % 2 == 1 {
+                if parity == 1 {
+                    odds.push(elem);
+                    parity = 0;
+                } else {
+                    evens.push(elem);
+                    parity = 1
+                }
+            }
+            *count = 0usize;
+        });
```

Comment:
This loop creates an iterator out of the HashMap `counters` which automatically performs a sort. To avoid the sort and preserve the ordering in `t` we iterate over `t` instead in the original loop.

Reply:
I agree that iterating over the map doesn't preserve the ordering of `t`, (and so we should iterate over `self.0` instead), but I don't agree that it sorts anything. The `HashMap` iteration order is arbitrary (determined by the hash values), not sorted. See: https://docs.rs/hashbrown/latest/hashbrown/hash_map/struct.HashMap.html#method.into_iter-2

Reply:
In this case though, I think `for` loop is preferable:

```rust
for elem in &self.0 {
    let count = counters.remove(elem).unwrap();
    let half_count = count / 2;
    evens.extend(vec![elem; half_count]);
    odds.extend(vec![elem; half_count]);
    if count % 2 == 1 {
        if parity == 1 {
            odds.push(elem);
            parity = 0;
        } else {
            evens.push(elem);
            parity = 1;
        }
    }
}
```

Reply:
Ah, you're right. It doesn't sort, it just changes the order.

Nice find with `IndexMap`! I'll change over to `IndexMap` and test with your original `for` loop.